support grpc protocol proxy (#793)
megaease/easegress: doc/cookbook/grpc-server.md

> # API Aggregation

Author: I forgot to commit the doc and now fixed it. If the content is not good yet, I will exclude the doc from this PR.

# gRRC Service Proxy

- [gRPC Service Proxy](#grpc-service-proxy)
    - [Similar to Service Proxy](#similar-to-service-proxy)
    - [Configure gRPC Filter](#configure gRPC Filter)

Easegress gRPC servers as a forward or reverse proxy base on gRPC protocol. 

## Similar to Service Proxy
Its design and implementation largely refer to the original HTTP Server, so I highly recommended that you read the [Service-Proxy documentation](./service-proxy.md) first.

### Configure Whether to Use Connection Pool

create a gRPC forward proxy filter to handler traffic

> notice: `connectTimeout` takes effect if and only when `blockUntilConnected` be specified as true

Reviewer: in this case, please merge `connectTimeout` and `blockUntilConnected` into one option: `connectTimeout`, when it has a valid value, we consider `blockUntilConnected` to be true, and if it is empty or zero (or other invalid value), we consider `blockUntilConnected` to be false, that's, `blockUntilConnected` can be removed.

``` yaml
filters:
  - kind: GRPCProxy
    pools:
      - loadBalance:
          # Using the forward load balancing strategy 
          policy: forward
          forwardKey: forward
        serviceName: "easegress-forward"
        blockUntilConnected: true
        connectTimeout: 300ms
    name: grpcforwardproxy
flow:
  - filter: grpcforwardproxy
    jumpIf: {}
kind: Pipeline
name: pipeline-grpc-forward
```

> tips: Some gRPC servers manage physical connections by themselves, and they do not consider gRPC gateway scenarios, such as Nacos: when the actual client goes offline, the server will detect and actively close the server's Connection. because gRPC is based on HTT2, it has the characteristics of request and response multiplexing, so there may be a connection from the same easegress used by multiple real clients to the real server. In this case, the server actively closes the connection, which will cause other normal clients to be affected.On the other hand, the server in grpc-go is designed to expect that business processing will not be aware of the underlying connection's status, so it does not provide related APIs for obtaining connection status. I have submitted a related [issue](https://github.com/grpc/grpc-go/issues/5835) in grpc-go to discuss that, you can check it out if you are interested.

Reviewer: ```suggestion
Note: Some gRPC servers manage physical connections by themselves, and they do not consider gRPC gateway scenarios, such as Nacos: when the actual client goes offline, the server will detect and actively close the server's Connection. because gRPC is based on HTTP2, it has the characteristics of request and response multiplexing, so there may be a connection from the same easegress used by multiple real clients to the real server. In this case, the server actively closes the connection, which will cause other normal clients to be affected. On the other hand, the server in grpc-go is designed to expect that business processing will not be aware of the underlying connection's status, so it does not provide related APIs for obtaining connection status. I have submitted a related [issue](https://github.com/grpc/grpc-go/issues/5835) in grpc-go to discuss that, you can check it out if you are interested.
```

# gRRC Service Proxy

- [gRPC Service Proxy](#grpc-service-proxy)
    - [Similar to Service Proxy](#similar-to-service-proxy)
    - [Configure Whether to Use Connection Pool](#configure whether to use connection pool)
       - [When Not to Use Connection Pool](#When-not-to-use-connection-pool)

Easegress gRPC servers as a forward or reverse proxy base on gRPC protocol. 

## Similar to Service Proxy

> Its design and implementation largely refer to the original HTTP Server, so it is strongly recommended that you read 
> the [Service-Proxy documentation](./service-proxy.md) first.

Reviewer 2: ```suggestion
Its design and implementation largely refer to the original HTTP Server, so it is strongly recommended that you read the [Service-Proxy documentation](./service-proxy.md) first.
```

### Configure Whether to Use Connection Pool

create a gRPC forward proxy filter to handler traffic

``` yaml
filters:
  - kind: GRPCProxy
    port: 8081
    useConnectionPool: true
    initConnNum: 2
    maxConnsPerHost: 1024
    connectTimeout: 200ms
    borrowTimeout: 1000ms
    pools:
      - loadBalance:
          #使用正向代理负载均衡策略
          policy: forward
          forwardKey: forward
        serviceName: "easegress-forward"
    name: grpcforwardproxy
flow:
  - filter: grpcforwardproxy
    jumpIf: {}
kind: Pipeline
name: pipeline-grpc-forward
```

> If you don't want to use connection pool , then set `useConnectionPool` to false or not specify. by the way,
> parameters such as `initConnNum`,`maxConnsPerHost`,`connectTimeout`,`borrowTimeout` are valid when `useConnectionPool` is
> true, otherwise, these parameters would be ignored.

Reviewer 2: ```suggestion
If you don't want to use the connection pool, then set `useConnectionPool` to false or not specified. by the way,parameters such as `initConnNum`, `maxConnsPerHost`, `connectTimeout`, `borrowTimeout` are valid when `useConnectionPool` is true, otherwise, these parameters would be ignored.
```

### When Not to Use Connection Pool

> Some gRPC servers manage physical connections by themselves, and they are not designed with gRPC gateway scenarios, such 
> as Nacos: when the actual client goes offline, the server will detect and actively close the server's Connection. 
> because gRPC is based on HTT2, it has the characteristics of request and response multiplexing, so there may be a 
> connection from the same easegress used by multiple real clients to the real server.In this case, the server actively 
> closes the connection, which will cause other normal clients to be affected.

Reviewer 2: ```suggestion
Some gRPC servers manage physical connections by themselves, and they are not designed with gRPC gateway scenarios, such as Nacos: when the actual client goes offline, the server will detect and actively close the server's Connection. because gRPC is based on HTT2, it has the characteristics of request and response multiplexing, so there may be a connection from the same easegress used by multiple real clients to the real server. In this case, the server actively closes the connection, which will cause other normal clients to be affected.
```
